tests/test_ray_map_reduce.py

```diff
@@ -0,0 +1,218 @@
+import unittest
+from datetime import timedelta
+from timeit import default_timer as timer
+from typing import List, Tuple
+
+from electionguard.elgamal import (
+    ElGamalKeyPair,
+    ElGamalCiphertext,
+    elgamal_encrypt,
+    elgamal_add,
+    elgamal_keypair_from_secret,
+)
+from electionguard.group import int_to_q, rand_q, ElementModQ
+from electionguard.nonces import Nonces
+from electionguard.utils import get_optional
+from electionguardtest.elgamal import elgamal_keypairs
+from hypothesis import settings, HealthCheck, given
+from hypothesis.strategies import lists, integers, text
+
+from arlo_e2e.eg_helpers import log_and_print
+from arlo_e2e.ray_helpers import ray_init_localhost
+from arlo_e2e.ray_map_reduce import MapReduceContext, RayMapReducer
+
+
+class StringLengthContext(MapReduceContext[str, int]):
+    def map(self, input: str) -> int:
+        return len(input)
+
+    def reduce(self, input: List[int]) -> int:
+        return sum(input)
+
+    def zero(self) -> int:
+        return 0
+
+
+class ElGamalEncryptor(MapReduceContext[Tuple[ElementModQ, int], ElGamalCiphertext]):
+    keypair: ElGamalKeyPair
+
+    def map(self, input: Tuple[ElementModQ, int]) -> ElGamalCiphertext:
+        nonce, plaintext = input
+        return get_optional(elgamal_encrypt(plaintext, nonce, self.keypair.public_key))
+
+    def reduce(self, input: List[ElGamalCiphertext]) -> ElGamalCiphertext:
+        return elgamal_add(*input)
+
+    def zero(self) -> ElGamalCiphertext:
+        nonce = rand_q()  # we'll get a different zero each time
+        return self.map((nonce, 0))
+
+    def __init__(self, keypair: ElGamalKeyPair):
+        self.keypair = keypair
+
+
+class TestRayMapReduce(unittest.TestCase):
```

Comment:
we should probably have tests for the corner cases, like 0 workers, 0 reduce shards, etc. to make sure the bounds assertions work. Also, check for inputs with only one item, etc. Though maybe we don't need to with hypothesis?

Reply:
Are we going for 100% code coverage? Because if we are, then there's a lot of extra stuff we need to test. Right now, we're floating somewhere around 80% overall.

Reply:
In general, yes. it might be good to at least get good coverage in PRs adding new features. @benadida may have more thoughts about this.

Reply:
yes I would like to get us to 100% test coverage for all things backend-y at least.